jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -503,130 +507,40 @@ public static String parentPath(String p)
      */
     public static String canonicalPath(String path)
     {
-        if (path==null || path.length()==0)
+        if (path == null || path.isEmpty() || !path.contains("."))
             return path;
 
-        int end=path.length();
-        int start = path.lastIndexOf('/', end);
+        if(path.startsWith("/.."))
+            return null;
 
-    search:
-        while (end>0)
-        {
-            switch(end-start)
-            {
-              case 2: // possible single dot
-                  if (path.charAt(start+1)!='.')
-                      break;
-                  break search;
-              case 3: // possible double dot
-                  if (path.charAt(start+1)!='.' || path.charAt(start+2)!='.')
-                      break;
-                  break search;
-            }
-            
-            end=start;
-            start=path.lastIndexOf('/',end-1);
-        }
+        List<String> directories = new LinkedList<>();
+        Collections.addAll(directories, path.split("(?<=\\/)"));
```

Comment:
One final change.... could you make this a static precompiled Pattern  eg
```` java
   static Pattern __PATH_SPLIT = Pattern.compile("(?<=\\/)");
````
and  then this line becomes the more efficient
````java
    Collections.addAll(directories, __PATH_SPLIT.split(path);
````

I'm pretty sure that Pattern is threadsafe in this way.